LOOP-1426: Remove untriggered delayed alerts that have been retracted

rickpasetto opened the pull request:
If a delayed alert is issued, but is retracted before it is "triggered" (i.e. before the delay expires), remove it from the database, under the theory that we only should record alerts shown to the user, not ones retracted before they are ever shown.

(Note: I just filed [LOOP-1426](https://tidepool.atlassian.net/browse/LOOP-1426), so we don't yet have consensus on the proper fix for this, nor is it in a sprint.  Nevertheless, I wanted to check the feasibility of the solution, and it was easier than I thought so I thought I'd put up this PR...)

(NOTE: this supercedes https://github.com/tidepool-org/Loop/pull/116 )

rickpasetto committed:
LOOP-1426: Remove untriggered delayed alerts that have been retracted

If a delayed alert is issued, but is retracted before it is "triggered" (i.e. before the delay expires), remove it from the database, under the theory that we only should record alerts shown to the user, not ones retracted before they are ever shown.

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -75,12 +75,23 @@ public class AlertStore {
     public func recordAcknowledgement(of identifier: Alert.Identifier, at date: Date = Date(),
                                       completion: ((Result<Void, Error>) -> Void)? = nil) {
-        recordUpdateOfLatest(of: identifier, addingPredicate: NSPredicate(format: "acknowledgedDate == nil"), with: { $0.acknowledgedDate = date }, completion: completion)
+        recordUpdateOfLatest(of: identifier, addingPredicate: NSPredicate(format: "acknowledgedDate == nil"),
```
[nit] Either all on only line or all on separate lines?

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -108,17 +119,21 @@ extension AlertStore {
 
     private func recordUpdateOfLatest(of identifier: Alert.Identifier,
```
[nit] `recordUpdateOfLatest` is getting more complicated to handle the different use cases for `recordAcknowledgement` and `recordRetraction`. To me, it is getting harder to reason about the common function than if they were just implemented separately. Consider implementing separately.

rickpasetto replied:
I'm sorry but I humbly disagree.  Separate functions would be largely duplicated code.  For example, it made it easier for you to add `purgeExpired`.  I'll try to make the code a little easier to read.

darinkrauss replied:
Your call.

darinkrauss reviewed (approved):
LGTM with a couple of nits for your consideration. 📌

rickpasetto committed:
executeQuery now excludes "future alerts"

From feedback from @ps2, querying the store should not return future alerts (i.e. delayed alerts whose interval goes beyond "now").
(the NSPredicate for this was a bit tricky, admittedly)

Also, includes some PR Feedback

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -72,17 +77,32 @@ public class AlertStore {
-        recordUpdateOfLatest(of: identifier, addingPredicate: NSPredicate(format: "acknowledgedDate == nil"), with: { $0.acknowledgedDate = date }, completion: completion)
+        recordUpdateOfLatest(of: identifier,
+                             addingPredicate: NSPredicate(format: "acknowledgedDate == nil"),
+                             with: { $0.acknowledgedDate = date; return .save },
```
[nit] I would put this on multiple lines like below, but your choice.

rickpasetto committed:
Added some clarifying comments.

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -218,11 +241,20 @@ extension AlertStore {
 
-    func executeQuery(since date: Date, limit: Int, completion: @escaping (QueryResult<SinceDateFilter>) -> Void) {
-        executeAlertQuery(from: QueryAnchor(filter: SinceDateFilter(date: date)), limit: limit, completion: completion)
+    func executeQuery(since date: Date, excludingFutureAlerts: Bool = true, now: Date = Date(), limit: Int, completion: @escaping (QueryResult<SinceDateFilter>) -> Void) {
```
[nit] Any reason you are swapping the order of these parameters (`now` and `excludingFutureAlerts`)?

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -218,11 +241,20 @@ extension AlertStore {
+        let excludingFutureAlerts: Bool
+        var predicate: NSPredicate? {
+            let datePredicate = NSPredicate(format: "issuedDate >= %@", date as NSDate)
+            // if case .delayed(let interval) = $0.trigger, $0.issuedDate + interval >= date {
```
[nit] Commented out?

darinkrauss reviewed (approved):
A couple nits, but LGTM. ☘️

rickpasetto committed:
Apply the "retraction removes untriggered alert" policy to repeating alerts too

More great feedback from Darin

rickpasetto committed:
More PR feedback (nits)

rickpasetto committed:
Use different predicate expressions depending on SQLite or in-memory

<grumble>

rickpasetto committed:
Fix comment

darinkrauss reviewed (approved):
LGTM. 🐻
